### PR TITLE
[chatgpt] Detached PPO Training

### PR DESCRIPTION
**Pull request by CsRic**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

'Detached' PPO training means that the experience makers and trainers are splitted into different nodes for asynchronous training. Models are not shared. 

1. Propose several classes for 'detached' manner: ExperienceMakerHolder, DetachedReplayBuffer, DetachedPPOTrainer
2. Implement Ray for Detached workflow structure
3. Examples: 1m1t. 1m2t, 2m1t, 2m2t.
4. Supported Strategies: Naive, DDP.
5. Won't affect present code.

Known issues:
~~1. Cannot detect cuda device on each worker.~~ fixed 20230324
2. Cannot run with Colossal strategy.
~~3. correctness of 1m1t.py example.~~ fixed 20230324

TODO:
- [x] Implement parameter update from trainer to experience maker
- [ ] fix issues
- [ ] Add TP strategy for experience maker and trainer.
- [ ] Support multiple nodes.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: applications/ChatGPT/examples/train_prompts.sh [hunk 1]
```diff
@@ -15,4 +15,7 @@ set_n_least_used_CUDA_VISIBLE_DEVICES() {
 
 set_n_least_used_CUDA_VISIBLE_DEVICES 2
 
-torchrun --standalone --nproc_per_node=2 train_prompts.py prompts.csv --strategy colossalai_zero2
+# torchrun --standalone --nproc_per_node=2 train_prompts.py prompts.csv --strategy colossalai_zero2
+
+torchrun --standalone --nproc_per_node=2 train_prompts.py "/home/lccsr/HPC-AI/ColossalAI/applications/ChatGPT/examples/awesome-chatgpt-prompts/prompts.csv" \
```
Comment (ht-zhou) [inline applications/ChatGPT/examples/train_prompts.sh:20]: better not to show the prompt path

File: applications/ChatGPT/chatgpt/experience_maker/detached.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        reward_model: \
+        initial_model: \
+        kl_coef:                 NaiveExperienceMaker init
+        experience_batch_size: batch size of generated experience
+    '''
+
+    def __init__(self, 
+                 detached_trainer_name_list: List[str], 
+                 actor: Actor,
+                 critic: nn.Module,
```
Comment (ht-zhou) [inline applications/ChatGPT/chatgpt/experience_maker/detached.py:27]: why to use nn.Module but not Critic?

File: applications/ChatGPT/chatgpt/replay_buffer/detached.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        tp_world_size: Number of workers in the same tp group
+        limit: Limit of number of experience sample BATCHs. A number <= 0 means unlimited. Defaults to 0.
+        cpu_offload: Whether to offload experience to cpu when sampling. Defaults to True.
+    '''
+
+    def __init__(self, sample_batch_size: int, tp_world_size: int = 1, limit : int = 0, cpu_offload: bool = True) -> None:
+        self.cpu_offload = cpu_offload
+        self.sample_batch_size = sample_batch_size
+        self.limit = limit
+        self.items = Queue(self.limit)
```
Comment (ht-zhou) [inline applications/ChatGPT/chatgpt/replay_buffer/detached.py:30]: what is the use of self.items ?

Comment (CsRic) [reply applications/ChatGPT/chatgpt/replay_buffer/detached.py:30]: It is the inner buffer, holding batched experience. 
I use ray.util.queue (torch.multiprocessing.queue is also ok) to deal with concurrent IO.

File: applications/ChatGPT/examples/train_prompts.sh (hunk 1 above)
Comment (CsRic) [reply applications/ChatGPT/examples/train_prompts.sh]: solved

File: applications/ChatGPT/chatgpt/experience_maker/detached.py (hunk 2 above)
Comment (CsRic) [reply applications/ChatGPT/chatgpt/experience_maker/detached.py]: solved

Comment (CsRic) [reply applications/ChatGPT/chatgpt/experience_maker/detached.py]: solved.

**Comment by binmakeswell** [comment]
Hi @CsRic Thanks for your contribution, but there is a conflict in this PR. Could you please solve them first? Thanks.
image

**Comment by CsRic** [comment]
> Can you move all files to `coati/ray`?

Done

File: applications/Chat/coati/experience_maker/strategy/base.py
Comment (ver217) [inline applications/Chat/coati/experience_maker/strategy/base.py:1]: Is this file essential?

Comment (CsRic) [reply applications/Chat/coati/experience_maker/strategy/base.py:1]: no.

File: applications/Chat/examples/train_dummy.sh [hunk 4]
```diff
@@ -14,5 +14,5 @@ set_n_least_used_CUDA_VISIBLE_DEVICES() {
 }
 
 set_n_least_used_CUDA_VISIBLE_DEVICES 2
-
+export CUDA_VISIBLE_DEVICES="0,1,2,3"
```
Comment (ver217) [inline applications/Chat/examples/train_dummy.sh:17]: remove this

File: applications/Chat/coati/ray/1m1t.py
Comment (ver217) [inline applications/Chat/coati/ray/1m1t.py:1]: Move examples to `examples/ray/`

File: applications/Chat/coati/experience_maker/strategy/base.py
Comment (ver217) [reply applications/Chat/coati/experience_maker/strategy/base.py:1]: Oh, please delete it.

**Review by ver217 — APPROVED** [review]

**Pull request merged**